### ISSUE TITLE
Miscompilation of `store undef, %ptr` (or an Alive2 bug?)

### ISSUE THREAD

**Issue by XChy** [llvm/llvm-project#86261]
Alive2 link: https://alive2.llvm.org/ce/z/-6JYLW

InstCombine folds:
```llvm
define void @src(ptr %p) {
  store i8 undef, ptr %p
  ret void
}
```
to
```llvm
define void @src(ptr %p) {
  ret void
}
```

But Alive2 told me that it's not a correct transform. AFAIK, the semantics of `undef` is like uninitialized value in memory. I'm not that sure whether it's a LLVM problem or an Alive2 problem, so post an upstream issue first.

**Comment by XChy** [comment]
cc @nunoplopes

**Comment by dtcxzyw** [comment]
cc @regehr

**Comment by dtcxzyw** [comment]
See the LangRef https://llvm.org/docs/LangRef.html#undefined-values
> A store of an undefined value can be assumed to not have any effect; we can assume that the value is overwritten with bits that happen to match what was already there. This argument is only valid if the stored value is provably not poison.

**Comment by nunoplopes** [comment]
Yep, that transformation is wrong because if there was a poison value at `%p`, replacing it with undef is not a no-op. Undef is less undefined than poison.

We haven't removed this transformation since no one benchmarked the impact + we wanted to remove undef altogether. Though there hasn't been a lot of work towards that direction lately.
